artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/ActiveMQPacketHandler.java

```diff
@@ -175,6 +171,11 @@ private void handleCreateSession(final CreateSessionMessage request) {
             activeMQPrincipal = connection.getDefaultActiveMQPrincipal();
          }
 
+         if (connection.getTransportConnection().getRedirectTo() != null) {
+            server.validateUser(activeMQPrincipal == null ? request.getUsername() : activeMQPrincipal.getUserName(), activeMQPrincipal == null ? request.getPassword() : activeMQPrincipal.getPassword(), connection, protocolManager.getSecurityDomain());
```

Comment:
Why not anticipating server.validateUser and passing the validatedUser parameter to createSession to avoid the double execution?

Reply:
that makes sense, create session is called in a few places, but preceding those with validate user is probably fine.
I think there is a body of work to deal with the sasl subject and probably replace a validatedUser string with a subject at some stage... but that is for another day.